controllers/vrg_kubeobjects.go

```diff
@@ -558,3 +663,283 @@ func kubeObjectsRequestsWatch(b *builder.Builder, kubeObjects kubeobjects.Reques
+			resourceName := resource[resourceType]
+
+			captureInstance, err := getResourceAndConvertToCaptureGroup(recipe, resourceType, resourceName)
+			if err != nil {
+				return resources, err
+			}
+
+			resources = append(resources, *captureInstance)
+		}
+	}
+
+	return resources, nil
+}
+
+func (v *VRGInstance) getRestoreGroupsFromWorkflow(
+	recipe Recipe.Recipe, workflow Recipe.Workflow) ([]ramen.KubeObjectsRecoverSpec, error,
+) {
+	resources := make([]ramen.KubeObjectsRecoverSpec, 0)
+
+	for _, resource := range workflow.Sequence {
+		// group: map[string]string, e.g. "group": "groupName", or "hook": "hookName"
+		for resourceType := range resource {
+			resourceName := resource[resourceType]
+
+			captureInstance, err := getResourceAndConvertToRecoverGroup(recipe, resourceType, resourceName)
+			if err != nil {
+				return resources, err
+			}
+
+			resources = append(resources, *captureInstance)
+		}
+	}
+
+	return resources, nil
+}
+
+func getResourceAndConvertToCaptureGroup(
+	recipe Recipe.Recipe, resourceType, name string) (*ramen.KubeObjectsCaptureSpec, error,
+) {
+	// check hooks OR groups
+	if resourceType == "group" {
+		for _, group := range recipe.Spec.Groups {
+			if group.Name == name {
+				return convertRecipeGroupToCaptureSpec(*group)
+			}
+		}
+
+		return nil, k8serrors.NewNotFound(schema.GroupResource{Resource: "Recipe.Spec.Group.Name"}, name)
+	}
+
+	if resourceType == "hook" {
+		hook, op, err := getHookAndOpFromRecipe(&recipe, name)
+		if err != nil {
+			return nil, k8serrors.NewNotFound(schema.GroupResource{Resource: "Recipe.Spec"}, resourceType)
+		}
+
+		return convertRecipeHookToCaptureSpec(*hook, *op)
+	}
+
+	return nil, k8serrors.NewNotFound(schema.GroupResource{Resource: "Recipe.Spec"}, resourceType)
+}
+
+// resource: could be Group or Hook
+func getResourceAndConvertToRecoverGroup(
+	recipe Recipe.Recipe, resourceType, name string) (*ramen.KubeObjectsRecoverSpec, error,
+) {
+	if resourceType == "group" {
+		for _, group := range recipe.Spec.Groups {
+			if group.Name == name {
+				return convertRecipeGroupToRecoverSpec(*group)
+			}
+		}
+
+		return nil, k8serrors.NewNotFound(schema.GroupResource{Resource: "Recipe.Spec.Group.Name"}, name)
+	}
+
+	if resourceType == "hook" {
+		hook, op, err := getHookAndOpFromRecipe(&recipe, name)
+		if err != nil {
+			return nil, k8serrors.NewNotFound(schema.GroupResource{Resource: "Recipe.Spec"}, resourceType)
+		}
+
+		return convertRecipeHookToRecoverSpec(*hook, *op)
+	}
+
+	return nil, k8serrors.NewNotFound(schema.GroupResource{Resource: "Recipe.Spec"}, resourceType)
+}
+
+func getHookAndOpFromRecipe(recipe *Recipe.Recipe, name string) (*Recipe.Hook, *Recipe.Operation, error) {
+	// hook can be made up of optionalPrefix/hookName; workflow sequence uses full name
+	var prefix string
+
+	var suffix string
+
+	const containsSingleDelim = 2
+
+	parts := strings.Split(name, "/")
+	switch len(parts) {
+	case 1:
+		prefix = ""
+		suffix = name
```

Comment:
This PR only contains "good path" usage handling, so duplicate names haven't been checked. Is this something we can (eventually) get the API server to check for us?

Reply:
I think an admission controller running on the API server could check for duplicates _before_ committing.

There is a `// +kubebuilder:validation:UniqueItems=true` marker, but I think it would check all fields in the structs, not just name.

Ideally, the groups, hooks and workflows would be maps instead of slices, each indexed by their name field.  In this way duplicate names would not be possible and the controller could access in o(1) instead of o(n).

However, sadly the API conventions [discourage it](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#lists-of-named-subobjects-preferred-over-maps) because a key looks like a field name in YAML.

Sever-side apply added the `//+listType=map` and the corresponding `//+listMapKey=` [markers](https://kubernetes.io/docs/reference/using-api/server-side-apply/#merge-strategy) for specifying a merge strategy.  I'd say there is a good chance they could be used to reject duplicate names, but I haven't tried them yet.

Reply:
After some discussion with @hatfieldbrian , the best solution to this was actually a change to the Recipe CRD, which I made a PR for [here](https://github.com/RamenDR/recipe/pull/2). The PR does not allow duplicate Names of Hooks, Groups or Workflows, and uses the `//+listType=map` and `//+listMapKey=name` markers specified above to do this.